Sources/AST/ASTVisitor/ASTVisitor.swift

```diff
@@ -511,6 +513,28 @@ public struct ASTVisitor {
                          passContext: postProcessResult.passContext)
   }
 
+  func visit(_ doCatchStatement: DoCatchStatement, passContext: ASTPassContext) -> ASTPassResult<DoCatchStatement> {
+    var passContext = passContext
+    var processResult = pass.process(doCatchStatement: doCatchStatement, passContext: passContext)
+
+    let scopeContext = passContext.scopeContext
+    processResult.element.doBody = processResult.element.doBody.map { statement in
+      return processResult.combining(visit(statement, passContext: processResult.passContext))
+    }
+
+    processResult.element.catchBody = processResult.element.catchBody.map { statement in
```

Comment:
Regarding the scoping

I'm not 100% on how this is done either but it seems like we call 
`processResult.passContext.scopeContext = scopeContext` to reset the scope context to the original context. This should happen after visiting the `doBody` and again after the `catchBody`. See the `IfStatement` for an example of what I mean. 

The TL;DR is that I think you need to put  `processResult.passContext.scopeContext = scopeContext` above this line in order to ensure that anything defined in the scope of the `do` is not in scope in the `catch` part. 

This is obviously something we don't want:

```swift
do {
    let x: 100
} catch is Bla {
    let y: x *2
}
```

Thank you for flagging this below.

Reply:
Updated according the TLDR. I moved the line up.